lib/plugins/aws/invokeLocal/index.js

```diff
@@ -300,30 +316,62 @@ class AwsInvokeLocal {
     );
   }
 
+  getDockerImageName() {
+    return `sls-docker-${this.getRuntime()}`;
+  }
+
   buildDockerImage(layerPaths) {
     const runtime = this.getRuntime();
-    const imageName = 'sls-docker';
+    const imageName = this.getDockerImageName();
 
-    return new BbPromise((resolve, reject) => {
-      let dockerfile = `FROM lambci/lambda:${runtime}`;
-      for (const layerPath of layerPaths) {
-        dockerfile += `\nADD --chown=sbx_user1051:495 ${layerPath} /opt`;
-      }
-      mkdirp.sync(path.join('.serverless', 'invokeLocal'));
-      const dockerfilePath = path.join('.serverless', 'invokeLocal', 'Dockerfile');
-      fs.writeFileSync(dockerfilePath, dockerfile);
-      this.serverless.cli.log('Building Docker image...');
-      const docker = spawn('docker', [
-        'build',
-        '-t',
-        imageName,
-        `${this.serverless.config.servicePath}`,
-        '-f',
-        dockerfilePath,
-      ]);
-      docker.on('exit', error => {
-        return error ? reject(error) : resolve(imageName);
+    let dockerfile = `FROM lambci/lambda:${runtime}`;
+    for (const layerPath of layerPaths) {
+      dockerfile += `\nADD --chown=sbx_user1051:495 ${layerPath} /opt`;
+    }
+
+    const dockerfilePath = path.join('.serverless', 'invokeLocal', runtime, 'Dockerfile');
+    const dockerfileCachePath = path.join(cachePath, 'dockerfiles', runtime, 'Dockerfile');
+
+    let cachedDockerfileEquals = BbPromise.resolve();
+
+    if (this.serverless.utils.fileExistsSync(dockerfileCachePath)) {
+      cachedDockerfileEquals = fs.readFileAsync(dockerfileCachePath).then(contents => {
+        return contents.toString() === dockerfile && this.checkDockerImage(imageName);
       });
+    }
+
+    return cachedDockerfileEquals.then(cacheEquals => {
+      if (cacheEquals) {
+        return imageName;
+      }
+      return BbPromise.all([
+        mkdirp.async(path.join('.serverless', 'invokeLocal', runtime)),
+        mkdirp.async(path.join(cachePath, 'dockerfiles', runtime)),
+      ])
+        .then(() => {
+          this.serverless.cli.log('Writing Dockerfile...');
+          return BbPromise.all([
+            fs.writeFileSync(dockerfilePath, dockerfile),
+            fs.writeFileSync(dockerfileCachePath, dockerfile),
+          ]);
+        })
+        .then(
+          () =>
+            new BbPromise((resolve, reject) => {
+              this.serverless.cli.log('Building Docker image...');
+              const docker = spawn('docker', [
+                'build',
+                '-t',
+                imageName,
+                `${this.serverless.config.servicePath}`,
+                '-f',
+                dockerfilePath,
+              ]);
+              docker.on('exit', error => {
+                return error ? reject('Failed to build docker image') : resolve(imageName);
```

Comment:
This will actually only return the exit code of the application rather than the error. The error will exist in `docker.stderr.on("data")` event, but I can return that buffered result :)

Reply:
In either case it should be error instance.

Therefore if argument as passed to exit is an exit code. Let's maybe do:

```javascript
docker.on('exit', exitCode => {
  return exitCode ? reject(new Error(`Failed to build docker image (exit code ${exitCode})`)) : ..
```